clog/clog.go

```diff
@@ -0,0 +1,196 @@
+)
+
+// Verbose is a boolean type that implements Infof (like Printf) etc.
+// See the documentation of V for more information.
+type Verbose bool
+
+var stdKeys map[string]bool
+var stdKeysOrder = []string{manifestID, sessionID, nonce, seqNo}
+
+func init() {
+	stdKeys = make(map[string]bool)
+	for _, key := range stdKeysOrder {
+		stdKeys[key] = true
+	}
+}
+
+func V(level glog.Level) Verbose {
+	return Verbose(bool(glog.V(level)))
+}
+
+type values struct {
+	mu   sync.RWMutex
+	vals map[string]string
+}
+
+func newValues() *values {
+	return &values{
+		vals: make(map[string]string),
+	}
+}
+
+// Clone creates new context with parentCtx as parent and
+// logging details from logCtx
+func Clone(parentCtx, logCtx context.Context) context.Context {
+	cmap, _ := logCtx.Value(clogContextKey).(*values)
+	newCmap := newValues()
+	if cmap != nil {
+		cmap.mu.RLock()
+		for k, v := range cmap.vals {
+			newCmap.vals[k] = v
+		}
+		cmap.mu.RUnlock()
+	}
+	return context.WithValue(parentCtx, clogContextKey, newCmap)
+}
+
+// Same creates new context with parentCtx as parent and
+// same logging data as in logCtx
+func Same(parentCtx, logCtx context.Context) context.Context {
+	cmap, _ := logCtx.Value(clogContextKey).(*values)
+	if cmap == nil {
+		cmap = newValues()
+	}
+	return context.WithValue(parentCtx, clogContextKey, cmap)
+}
+
+func AddManifestID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, manifestID, val)
+}
+
+func AddOrchSessionID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, orchSessionID, val)
+}
+
+func AddSessionID(ctx context.Context, val string) context.Context {
+	return AddVal(ctx, sessionID, val)
+}
+
+func AddNonce(ctx context.Context, val uint64) context.Context {
+	return AddVal(ctx, nonce, strconv.FormatUint(val, 10))
+}
+
+func AddSeqNo(ctx context.Context, val uint64) context.Context {
+	return AddVal(ctx, seqNo, strconv.FormatUint(val, 10))
+}
+
+func AddVal(ctx context.Context, key, val string) context.Context {
+	cmap, _ := ctx.Value(clogContextKey).(*values)
+	if cmap == nil {
+		cmap = newValues()
+		ctx = context.WithValue(ctx, clogContextKey, cmap)
+	}
+	cmap.mu.Lock()
+	cmap.vals[key] = val
+	cmap.mu.Unlock()
+	return ctx
+}
+
+func Warningf(ctx context.Context, format string, args ...interface{}) {
+	glog.WarningDepth(1, formatMessage(ctx, false, format, args...))
+}
+
+func Errorf(ctx context.Context, format string, args ...interface{}) {
+	glog.ErrorDepth(1, formatMessage(ctx, false, format, args...))
+}
+
+func Fatalf(ctx context.Context, format string, args ...interface{}) {
+	glog.FatalDepth(1, formatMessage(ctx, false, format, args...))
+}
+
+func Infof(ctx context.Context, format string, args ...interface{}) {
+	infof(ctx, false, format, args...)
+}
+
+// Infofe if last argument is not nil it will be printed as " err=%q"
+func Infofe(ctx context.Context, format string, args ...interface{}) {
```

Comment:
It is convention that I've just invented 😄 
This exists to avoid having `err=<nil>` in the logs. We will get it if we're doing `Infof("err=%v", err)` in case `err` could be `nil`.
So to avoid this I need to write ```golang
if err != nil {
  glog.Infof("message var1=%s err=%v", var1, err)
} else {
  glog.Infof("message var1=%s", var1)
}`
`Infofe` helps to avoid this duplication.

Reply:
ok, I see. I'd probably prefer having `err=<nil>` since it's kind-of magic behavior of the library you create. In other words, as a user of `clog` you can be confused that suddenly ` err=` is automatically added at the end. Anyway, it's minor and probably opinion-based, so I leave it up to you (and @victorges) to decide.

Reply:
You would prefer not to have `err=<nil>` in the logs once you will start digging through actual logs 😄 
When searching through logs using `err=` as a query it's much better to see lines with actual errors 😄

Reply:
> When searching through logs using err= as a query it's much better to see lines with actual errors 😄

This will be so much less insane now that we have an actual tool for querying logs by their actual fields and not just text search 😄 
But I think it's nice to avoid the `err=<nil>` lines anyway :)

> It is convention that I've just invented 😄

Since it is a new convetion and thus not something that is self-explanatory, WDYT of using a more explicit name for the function? I like `Infoferr` for example. 

And another approach is providing an `AddErr` function that only adds the error to the log if it is not `nil`. On your example, it could be:
```golang
ctx = clog.AddErr(ctx, err)
clog.Infof(ctx, "message var1=%s", var1)
```
or with the chaining idea in the other comment:
```golang
clog.Fields(ctx).AddErr(err)
    .Infof("message var1=%s", var1)
```